packages/manager/modules/pci/src/projects/project/storages/databases/database/users/add/add.controller.js

```diff
@@ -29,12 +34,105 @@ export default class AddUserCtrl {
     if (this.isRolesReadOnly) {
       this.model.selectedRoles = [];
     }
+    if (this.isAdvancedRole) {
+      this.addEmptyEntry();
+    }
+  }
+
+  isButtonAddDisplayed($index) {
+    return !this.model.selectedRoles[$index].added;
+  }
+
+  isButtonDeleteDisplayed($index) {
+    const isFirstItem = $index === 0 && !this.model.selectedRoles[$index].added;
+    return this.model.selectedRoles[$index].added || isFirstItem;
+  }
+
+  isButtonAddEnabled($index) {
+    const isFilled =
+      this.model.selectedRoles[$index].key &&
+      (this.model.selectedRoles[$index].key.admin ||
+        this.model.selectedRoles[$index].value);
+    return isFilled && !this.processing;
+  }
+
+  isButtonDeleteEnabled($index) {
+    const isFirstItem = $index === 0 && !this.model.selectedRoles[$index].added;
+    return (
+      isFirstItem ||
+      (this.model.selectedRoles[$index].added && !this.processing)
+    );
+  }
+
+  addEmptyEntry() {
+    this.getAvailableRoles();
+    if (this.availableRoles.length > 0) {
+      this.model.selectedRoles.push({});
+    }
+  }
+
+  onAdvancedRoleChanged(modelValue, $index) {
+    this.model.selectedRoles[$index].key = null;
+    // reset value on field change to avoid invalid data
+    this.model.selectedRoles[$index].value = undefined;
+    this.$timeout(() => {
+      this.model.selectedRoles[$index].key = modelValue;
+    });
+  }
+
+  onAddAdvancedRole($index) {
+    const addedRole = this.model.selectedRoles[$index];
+    // check error
+    if (
+      !addedRole.admin &&
+      this.model.selectedRoles.filter(
+        (role) =>
+          role.key.name === addedRole.key.name &&
+          role.value === addedRole.value,
+      ).length > 1
```

Comment:
**nitpick:** `Array.prototype.some` is more suitable since boolean is expected and the loop can be terminated after finding the first match.
```suggestion
      this.model.selectedRoles.some(
        (role) =>
          role.key.name === addedRole.key.name &&
          role.value === addedRole.value,
      )
```